feat: Add more state props from controllers to Sentry mask

pedronfigueiredo opened the pull request:
## Explanation

Before sending any logs to our error-monitoring system, we filter all PII out of the app state.

However, we have been leaving a lot of important state data out of the error monitoring systems, which sometimes complicates debugging unnecessarily.

Building on @Gudahtt's work, who previously split `SENTRY_BACKGROUND_STATE` on #20426, this PR adds more props to our explicit error monitoring masking policy using that constant.

This was achieved by manually checking each controller state following the trail from the following section of `metamask-controller.js`:

https://github.com/MetaMask/metamask-extension/blob/9514b47a7eb8486584741d69fd26f8a52a3de10a/app/scripts/metamask-controller.js#L1636-L1679

### Snapshotting state properties delta

Given the number of controllers and corresponding state properties in our app, we want to have a clear view of which properties we intentionally chose to leave out of the policy, if any. As a reminder, properties that are not on the policy are NOT sent to our error monitoring systems. They are equivalent to being set to `false` in our `SENTRY_BACKGROUND_STATE` constant.

But even more importantly, we want to surface opportunities to enrich our error logs to developers who introduce state properties to the extension in any future PRs. On #20514, @Gudahtt first introduced snapshot support for e2e tests. On this PR, we leverage this pattern to capture the delta between what's explicitly set on the `SENTRY_BACKGROUND_STATE` constant and the background process state tree that is accessible in the UI -- `uiState`.

Fixes #20575

pedronfigueiredo committed:
feat: update sentry mask adding  controller props to improve error monitoring

pedronfigueiredo commented on test/e2e/tests/errors.spec.js:
```diff
@@ -737,3 +744,92 @@ describe('Sentry errors', function () {
   });
 });
+
+describe.only('Sentry App State Mask', function () {
```
I will extract this to a separate file, and put the `matchesSnapshot` function in a shared file.

pedronfigueiredo replied:
remove the `.only`

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
+  DesktopController: {
+    desktopEnabled: true,
+  },
+  // we're not sending any data to Sentry from the following controllers
```
It should be safe to set all controllers to `{}`, as that will just show us what the top-level controller properties are

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/db9e2566e13811c8ddb6c590cc4370f8e74163f9

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -30,6 +30,53 @@ export const SENTRY_BACKGROUND_STATE = {
   },
+  PermissionLogController: {
+    // double-check if contains PID?
+    permissionHistory: true,
```
These entries would have origins, so maybe capturing the types is enough for now. If we want more insights from this, we could update the mask function to mask dynamic collections (arrays/objects), which isn't supported yet

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/b3330828235d61e68ab356e3e859a30d3c44af6f

Gudahtt commented:
Great work! I'll review the changes, I've only gotten a few properties in so far.

The e2e tests are failing because we need an errors state snapshot update due to the new property. You can fix it by running `yarn test:e2e:single --browser chrome test/e2e/tests/errors.spec.js --update-snapshot` locally (after creating a test build, e.g. `yarn start:test`  for a quick build).

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -30,6 +30,53 @@ export const SENTRY_BACKGROUND_STATE = {
+  },
+  TokenListController: {
+    // double-check if contains PID?
+    tokenList: true,
```
This one is borderline, I don't think we should include it in its entirety at least. Probably safest to exclude for now. Same for `tokensChainsCache`

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/b3330828235d61e68ab356e3e859a30d3c44af6f

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -30,6 +30,53 @@ export const SENTRY_BACKGROUND_STATE = {
+  },
+  TokensController: {
+    // double-check if contains PID?
+    tokens: true,
```
I'd advice omitting all of these as well, just keeping `{}` for the TokensController so that we can see the property types in the mask

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/b3330828235d61e68ab356e3e859a30d3c44af6f

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -30,6 +30,53 @@ export const SENTRY_BACKGROUND_STATE = {
+  },
+  NotificationController: {
+    // double-check if contains PID?
+    notifications: true,
```
This is part of snaps, I'm not sure exactly what it contains. Maybe best to exclude it for now, especially since it won't shed light on v10.35.0 issues on stable anyway (no snaps there).

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/b3330828235d61e68ab356e3e859a30d3c44af6f

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -30,6 +30,53 @@ export const SENTRY_BACKGROUND_STATE = {
+    allDetectedTokens: true,
+  },
+  SmartTransactionsController: {
+    smartTransactions: false,
```
This will include full transaction objects, so we'll want to exclude this

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
+      ticker: true,
+      nickname: true,
+      rpcPrefs: false,
+      id: false,
```
This is a UUID that might be referenced elsewhere in state eventually, so it seems safe to include. It wouldn't be useful today, but it might be later.

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/b3330828235d61e68ab356e3e859a30d3c44af6f

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
-    currentLocale: true,
     featureFlags: true,
-    forgottenPassword: true,
+    // double-check if contains PID?
```
Probably best to exclude this. It might contain a secret (the only method we support today of using private ipfs servers is to embed the secret in the URL)

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/db9e2566e13811c8ddb6c590cc4370f8e74163f9

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
+    advancedGasFee: true,
+    incomingTransactionsPreferences: true,
+    // double-check if contains PID?
+    knownMethodData: true,
```
Maybe better to omit this. It would hint at which transactions the user had encountered

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/db9e2566e13811c8ddb6c590cc4370f8e74163f9

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
   },
+  SwapsController: {
+    swapsState: {
+      // double-check if contains PID?
```
I think it would reveal which swap they were getting quotes for, so better to exclude for now

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/db9e2566e13811c8ddb6c590cc4370f8e74163f9

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
+      quotesPollingLimitEnabled: true,
+      fetchParams: true,
+      // double-check if contains PID?
+      tokens: true,
```
Hmm, not entirely sure if this would be PII or not (probably not, since it should be the full token list our API returns?) but it would make sense to exclude just to reduce the size of this snapshot. It can get pretty large.

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/db9e2566e13811c8ddb6c590cc4370f8e74163f9

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
+  },
+  SnapController: {
+    // double-check if contains PID?
+    snaps: true,
```
The snaps team has marked these as non anonynimized: https://github.com/MetaMask/snaps/blob/315276782b90269982e02f351b73f182c716d738/packages/snaps-controllers/src/snaps/SnapController.ts#L691

So better to exclude these. Likewise for the CronjobController: https://github.com/MetaMask/snaps/blob/315276782b90269982e02f351b73f182c716d738/packages/snaps-controllers/src/cronjob/CronjobController.ts#L91

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/db9e2566e13811c8ddb6c590cc4370f8e74163f9

Gudahtt commented on test/e2e/tests/errors.spec.js:
```diff
@@ -737,3 +744,92 @@ describe('Sentry errors', function () {
+          window.stateHooks?.getCleanAppState?.(),
+        );
+
+        const missingMetaMaskStatePropsInSentryMask = compareStates(
+          uiState,
+          SENTRY_BACKGROUND_STATE,
+        );
```
We have masks for the UI and background state. Since we're comparing against the UI state here, we should probably use the UI mask as the point of comparison. It includes the flattened background state in it already.

For the comparison function, maybe we could create an inverse mask function instead? That'd give us an object representing everything the mask excluded. e.g.


```suggestion
export function maskObject(object, mask) {
  return Object.keys(object).reduce((state, key) => {
    if (mask[key] === true) {
      state[key] = object[key];
    } else if (mask[key]) {
      state[key] = maskObject(object[key], mask[key]);
    } else {
      state[key] = object[key];
    }
    return state;
  }, {});
}
```

(it's a copy of `maskObject` except the true and falsy cases are reversed

Gudahtt replied:
I didn't use this idea after, but I did update this to compare with the UI mask in https://github.com/MetaMask/metamask-extension/pull/20595/commits/8b192813eb86a65175b8717b1835b320bc2b2c9b

Gudahtt commented on test/e2e/tests/state-snapshots/sentry-mask-policy-gaps.json:
```diff
@@ -0,0 +1,130 @@
+  "previousUserTraits.undefined",
+  "smartTransactionsState.liveness",
+  "snapsInstallPrivacyWarningShown",
+  "termsOfUseLastAgreed",
```
Seems to be missing from the app-state controller

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/b9210b7553f2db3b7dbb53c3da9bf7f71cf7ee14

Gudahtt commented on test/e2e/tests/state-snapshots/sentry-mask-policy-gaps.json:
```diff
@@ -0,0 +1,130 @@
+  "networksMetadata.networkConfigurationId.EIPS.1559",
+  "networksMetadata.networkConfigurationId.status",
+  "nextNonce",
+  "pendingApprovalCount",
```
This would be nice to include (it's in the ApprovalController)

Gudahtt replied:
That's the correct place, good catch. I'm not sure why that's marked as `anonymous: false`. Sometimes that setting is overlooked or misunderstood - it's easy for that to happen because it's not used for anything functional yet.

Leaving it out to stay on the safe side would be reasonable though.

DDDDDanica committed:
fix:remove changes in chrome-driver dependency

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
   MetaMetricsController: {
     metaMetricsId: true,
     participateInMetaMetrics: true,
+    eventsBeforeMetricsOptIn: true,
```
These queued events might allow linking a metrics ID to a Sentry report, which isn't something we're intentionally doing right now so I'll omit it for the time being.

Gudahtt replied:
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/b3330828235d61e68ab356e3e859a30d3c44af6f

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -64,36 +129,141 @@ export const SENTRY_BACKGROUND_STATE = {
+    snapsAddSnapAccountModalDismissed: true,
   },
   SignatureController: {
+    unapprovedMsgs: false,
```
These would reveal pending signatures, which isn't something we otherwise expose. These can be omitted.

Gudahtt committed:
Remove properties from mask

Gudahtt committed:
Add more values to mask

Gudahtt committed:
Sort the mask alphabetically

Gudahtt commented on package.json:
```diff
@@ -445,7 +445,7 @@
     "chalk": "^4.1.2",
     "chokidar": "^3.5.3",
-    "chromedriver": "^114.0.0",
+    "chromedriver": "^116.0.0",
```
Resolved in https://github.com/MetaMask/metamask-extension/pull/20595/commits/3f348a7c7fce261513744af29b516cde4a13e57b

Gudahtt committed:
Add termsOfUseLastAgreed to mask

Gudahtt committed:
Fix test imports

Gudahtt committed:
Update policy gap test to compare UI mask

Gudahtt committed:
Reorganize tests under one describe block

Gudahtt committed:
Update snapshots

Gudahtt commented:
Bringing this out of draft because I think it's in good enough shape, but I'll be giving this another review pass tomorrow

Gudahtt committed:
Mask another timestamp in state snapshots

Gudahtt committed:
Mask browser environment properties

Gudahtt committed:
Add missing UI field mask, and refactor field masking/removal

Gudahtt committed:
Eliminate remaining policy gaps

Gudahtt committed:
Simplify ganache options

Gudahtt committed:
Eliminate extra mask properties

Gudahtt committed:
Update mask to capture dynamic keys

The mask now supports dynamic keys. This lets set more fine-grained rules
for which data to include within dynamic data structures.

The mask has been updated to include just top-level keys for various
token-related data collections in state. This lets us see the chain IDs
that users have tokens on. This will be useful in debugging Sentry
reports of invalid keys in these data structures.

Gudahtt committed:
Add additional 'expected missing state' entries

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -28,72 +29,269 @@ export const ERROR_URL_ALLOWLIST = {
-    networkStatus: true,
+    networksMetadata: true,
     providerConfig: {
+      chainId: false,
```
Maybe this should be included? We're a bit inconsistent here, including chain IDs in some cases but not others. We can always re-evaluate later though.

Gudahtt replied:
```suggestion
      chainId: true,
```

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -28,72 +29,269 @@ export const ERROR_URL_ALLOWLIST = {
+  SwapsController: {
+    swapsState: {
+      approveTxId: false,
+      customApproveTxData: true,
```
This data likely just contains a custom approval limit, but I'll exclude it for now because I'm not 100% certain

```suggestion
      customApproveTxData: false,
```

Gudahtt commented on app/scripts/lib/setupSentry.js:
```diff
@@ -28,72 +29,269 @@ export const ERROR_URL_ALLOWLIST = {
+    web3ShimUsageOrigins: false,
+  },
+  AnnouncementController: {
+    announcements: true,
```
```suggestion
    announcements: false,
```

Gudahtt committed:
Remove unnecessary properties from state snapshot

Gudahtt committed:
Add providerConfig.chainId to state snapshot

Gudahtt committed:
Update error state snapshots
